[GTK][WPE] Prefer 8 bit DMA-BUF formats

carlosgcampos opened the pull request:
#### 5a3c1c46999cad53a9bcc11b8e25b2ca2c0567fe
<pre>
[GTK][WPE] Prefer 8 bit DMA-BUF formats
<a href="https://bugs.webkit.org/show_bug.cgi?id=267446">https://bugs.webkit.org/show_bug.cgi?id=267446</a>

Reviewed by Adrian Perez de Castro.

I thought the formats returned by the compositor were sorted by
preference, but only the tranches are sorted by usage, so all formats
for the same usage has the same priority. Some, drivers are returning 10
bits formats first, but we should prefer 8 bits ones if supported.

* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::dmabufFormats): Sort supported formats and
iterate them to build the list checking if they are available.
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::setupBufferFormat): Split
the format list by usage and iterate the supported list finding in every
usage list.

Canonical link: <a href="https://commits.webkit.org/272962@main">https://commits.webkit.org/272962@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ec39d56cc7c3c68222b294df5e4b3be50c9f3df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33745 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12520 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35686 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36363 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/30598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34795 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14895 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9672 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/29673 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34224 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/10564 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/30090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9215 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/30107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37685 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/30633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/30428 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/35439 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9453 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/7393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/33329 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11232 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10035 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4347 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10238 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/graphics/PlatformDisplay.cpp:
```diff
@@ -576,16 +576,16 @@ const Vector<PlatformDisplay::DMABufFormat>& PlatformDisplay::dmabufFormats()
-        static const Vector<uint32_t> s_supportedFormats = {
+        static const Vector<EGLint> s_supportedFormats = {
+            DRM_FORMAT_ARGB8888, DRM_FORMAT_XRGB8888, DRM_FORMAT_ABGR8888, DRM_FORMAT_XBGR8888,
+            DRM_FORMAT_RGBA8888, DRM_FORMAT_RGBX8888, DRM_FORMAT_BGRA8888, DRM_FORMAT_BGRX8888,
```
I wonder if it would make sense to list all the formats with alpha first, and then the ones without: ARGB > ABGR > RGBA > BGRA >> XRGB > XBGR > RGBX > XRGB.

It looks to me that in general we may want to increase the chances of choosing a format with alpha, in case e.g. web views are configured with a non-opaque background color. Or, even better: prefer XRGB variants when the background is opaque, ARGB otherwise... but this would be quite some more work and IMO out of the scope for this one patch 😉

aperezdc reviewed (approved):
The change to have into account that the list of formats is sorted by what WebKit prefers is good. I have only a small comment but not a blocker for landing.

carlosgcampos committed:
[GTK][WPE] Prefer 8 bit DMA-BUF formats
https://bugs.webkit.org/show_bug.cgi?id=267446

Reviewed by Adrian Perez de Castro.

I thought the formats returned by the compositor were sorted by
preference, but only the tranches are sorted by usage, so all formats
for the same usage has the same priority. Some, drivers are returning 10
bits formats first, but we should prefer 8 bits ones if supported.

* Source/WebCore/platform/graphics/PlatformDisplay.cpp:
(WebCore::PlatformDisplay::dmabufFormats): Sort supported formats and
iterate them to build the list checking if they are available.
* Source/WebKit/WebProcess/WebPage/dmabuf/AcceleratedSurfaceDMABuf.cpp:
(WebKit::AcceleratedSurfaceDMABuf::SwapChain::setupBufferFormat): Split
the format list by usage and iterate the supported list finding in every
usage list.

Canonical link: https://commits.webkit.org/272962@main
